### ISSUE TITLE
Difficulty reproducing zero-shot findings classification F1

### ISSUE THREAD

**Issue by shivi-ka** [StanfordMIMI/Merlin#34]
I am trying to reproduce the **zero-shot findings classification results** reported in the paper, but I am consistently observing a significant gap in F1 score. I suspect I may be missing some details in the prediction or evaluation pipeline and would really appreciate your guidance.

---

##  My Current Pipeline (Internal Merlin Test Set)

1. **Training**

   * Trained a Merlin model from scratch using CT images (no EHR) + Text (with splitting)
   * Selected checkpoint from **epoch 154** (best validation loss, batch size = 8)

2. **Zero-shot Prediction**

   * Used [disease_prompts.py](https://github.com/shivi-ka/merlin-assets/blob/main/disease_prompts.py)  for positive/negative prompts
   * Computed cosine similarity between CT embeddings and Prompt embeddings
   * Prediction rule is `1` if `pos_score > neg_score`  else `0`

3. **Evaluation**

   * Ground truth: `zero_shot_findings_disease_cls.csv` (mined labels)
   * Test split: derived from `reports_final.xlsx`
   * Ignored `-1` labels during F1 computation
   * Script used [evaluate_zeroshot_findings_self154.py](https://github.com/shivi-ka/merlin-assets/blob/main/evaluate_zeroshot_findings_self154.py)

```
===== F1 SCORE PER FINDING =====
submucosal_edema               -> 0.6559
renal_hypodensities            -> 0.2547
aortic_valve_calcification     -> 0.8933
coronary_calcification         -> 0.4706
thrombosis                     -> 0.7347
metastatic_disease             -> 0.5065
pancreatic_atrophy             -> 0.1928
renal_cyst                     -> 0.4642
osteopenia                     -> 0.6416
surgically_absent_gallbladder  -> 0.8364
atelectasis                    -> 0.6277
aortic_aneurysm                -> 0.4722
anasarca                       -> 0.5259
hernia                         -> 0.2974
lymphadenopathy                -> 0.4430
prostatomegaly                 -> 0.1596
biliary_ductal_dilation        -> 0.4966
cardiomegaly                   -> 0.5706
splenomegaly                   -> 0.2271
hepatomegaly                   -> 0.3359
atherosclerosis                -> 0.9495
ascites                        -> 0.9277
pleural_effusion               -> 0.9001
hepatic_steatosis              -> 0.6961
appendicitis                   -> 0.6136
gallstones                     -> 0.4930
hydronephrosis                 -> 0.1384
bowel_obstruction              -> 0.2075
free_air                       -> 0.1682
fracture                       -> 0.6082

Mean over findings F1: 0.5170
```

Paper report this F1 Score 0.730 

---

##  Ablation (Using Merlin Checkpoint)

- Use the same approach as above but evaluated using the released checkpoint: `i3_resnet_clinical_longformer_best_clip_04-02-2024_23-21-36_epoch_99.pt`
- Script used [evaluate_zeroshot_findings_merlin-i3d.py](https://github.com/shivi-ka/merlin-assets/blob/main/evaluate_zeroshot_findings_merlin-i3d.py)

```
===== F1 SCORE PER FINDING =====
submucosal_edema               -> 0.6457
renal_hypodensities            -> 0.2727
aortic_valve_calcification     -> 0.8322
coronary_calcification         -> 0.5055
thrombosis                     -> 0.7843
metastatic_disease             -> 0.5821
pancreatic_atrophy             -> 0.2281
renal_cyst                     -> 0.5370
osteopenia                     -> 0.6048
surgically_absent_gallbladder  -> 0.9175
atelectasis                    -> 0.7098
aortic_aneurysm                -> 0.4179
anasarca                       -> 0.7183
hernia                         -> 0.4335
lymphadenopathy                -> 0.4932
prostatomegaly                 -> 0.1881
biliary_ductal_dilation        -> 0.5290
cardiomegaly                   -> 0.6583
splenomegaly                   -> 0.2422
hepatomegaly                   -> 0.2013
atherosclerosis                -> 0.9631
ascites                        -> 0.8729
pleural_effusion               -> 0.8633
hepatic_steatosis              -> 0.8724
appendicitis                   -> 0.6400
gallstones                     -> 0.6530
hydronephrosis                 -> 0.1881
bowel_obstruction              -> 0.1969
free_air                       -> 0.1886
fracture                       -> 0.6759

Mean over findings F1: 0.5539
```

Paper reports this F1 Score 0.741

---

## Questions

I would really appreciate clarification on the following points:

1. Are the reported F1 results obtained from the released checkpoint `i3_resnet_clinical_longformer_best_clip_04-02-2024_23-21-36_epoch_99.pt`?

2. Can you release your zero short findinfs classification script that was use reported F1 scores?

3.  Can you confirm that `zero_shot_findings_disease_cls.csv` is the exact label file used for evaluation in the paper?

---

Any pointers on what I might be missing would be extremely helpful.

Thanks a lot for your time!

**Comment by ashwinkumargb** [comment]
Hi @shivi-ka , 

1. Yes, the reported results were obtained using the released checkpoint.

I’ve attached the zero-shot evaluation code in a zip file, including the scripts used for metric calculation. These should clarify how F1 scores are computed in our implementation.

Our internal development codebase is fairly complex, but I’ve included the core scripts needed to run the main components. They rely on standard implementations commonly used for multimodal training.

Regarding (3), the labels were identified using regex-based matching with specific positive and negative prompts. However, during training we used the zero_shot_utils.py file to handle label processing.

If you are observing low F1 scores, they may stem from differences in preprocessing or data loading, as the overall training framework (e.g., CLIP-style loss) follows standard practices.

I hope these materials are helpful. Please feel free to reach out if you have any further questions.

[Merlin_Sample_Code.zip](https://github.com/user-attachments/files/26225152/Merlin_Sample_Code.zip)

Regards,
Ashwin
